src/math/trig.c

```diff
@@ -9,568 +9,115 @@
+ *	+----+-----+-------+----+----+-------+--------+--------+
+ *	| 32 | 28  |  1	   | 32	| 31 |	 1   | 4.28   | 1.31   |
+ *	+------------------+-----------------+--------+--------+
+ */
 
-/* Sine lookup table read */
-static inline int32_t sine_lookup(int idx)
+int32_t sin_fixed(int32_t th_rad_fxp)
 {
-	int32_t s;
-	int i1;
+	/*cordic_atan2_lookup_table = atan(2.^-(0:N-1)) N = 31
+	 *CORDIC Gain is cordic_gain = prod(sqrt(1 + 2.^(-2*(0:31-1))))
+	 *Inverse CORDIC Gain,inverse_cordic_gain = 1 / cordic_gain
+	 */
+	static const int32_t cordicsine_lookup[31] = { 843314857, 497837829, 263043837, 133525159,
+	 67021687, 33543516, 16775851, 8388437, 4194283, 2097149, 1048576, 524288,
+	 262144, 131072, 65536, 32768, 16384, 8192, 4096, 2048, 1024, 512, 256, 128,
+	 64, 32, 16, 8, 4, 2, 1 };
+	int32_t b_idx;
+	int32_t b_yn;
+	int32_t xn;
+	int32_t xtmp;
+	int32_t ytmp;
+	int32_t cdc_sin_th;
+	int sign;
+	/*1073741823, deg = +57.2958*/
+	const int32_t CORDICSINE_P57DEG_Q30FL = Q_CONVERT_FLOAT(1.99999999813735, 29.9999999986564);
+	/* 652032874 , deg = 69.586061*/
+	const int32_t CORDICSINE_LUT_Q29FL  =  Q_CONVERT_FLOAT(1.214505869895220, 29);
+	/* 421657428 , deg = 90.000000 */
+	const int32_t CORDICSINE_PIOVERTWO_Q28FL  = Q_CONVERT_FLOAT(M_PI / 2, 28);
+	/* 843314857,  deg = 90.000000  */
+	const int32_t CORDICSINE_PIOVERTWO_Q29FL  = Q_CONVERT_FLOAT(M_PI / 2, 29);
+	/* 1686629713, deg = 90.000000  */
+	const int32_t CORDICSINE_PIOVERTWO_Q30FL  = Q_CONVERT_FLOAT(M_PI / 2, 30);
+	/*-1073741824, deg = -57.2958*/
+	/* Minor-Bug,-1073741824 = round(-1*2^30) but QCF is -1073741823*/
+	const int32_t CORDICSINE_N57DEG_Q30FL = Q_CONVERT_FLOAT(-1, 30);
 
-	i1 = idx & (2 * SINE_NQUART - 1);
-	if (i1 > SINE_NQUART)
-		i1 = 2 * SINE_NQUART - i1;
+	/*Addition or subtraction by a multiple of pi/2 is done in the data type
+	 *of the input. When the fraction length is 29, then the quantization error
+	 *introduced by the addition or subtraction of pi/2 is done with 29 bits of
+	 *precision.Input range of cordicsin must be in the range [-2*pi, 2*pi),
+	 *a signed type with fractionLength = wordLength-4 will fit this range
+	 *without overflow.Increase of fractionLength makes the addition or
+	 *subtraction of a multiple of pi/2 more precise
+	 */
+	if (th_rad_fxp > CORDICSINE_PIOVERTWO_Q28FL) {
+		if ((th_rad_fxp - CORDICSINE_PIOVERTWO_Q29FL) <= CORDICSINE_PIOVERTWO_Q28FL) {
+			th_rad_fxp -= CORDICSINE_PIOVERTWO_Q29FL;
+			sign  = -1;
+		} else {
+			th_rad_fxp -= CORDICSINE_PIOVERTWO_Q30FL;
+			sign  = 1;
+		}
+	} else if (th_rad_fxp < -CORDICSINE_PIOVERTWO_Q28FL) {
+		if ((th_rad_fxp + CORDICSINE_PIOVERTWO_Q29FL) >= -CORDICSINE_PIOVERTWO_Q28FL) {
+			th_rad_fxp += CORDICSINE_PIOVERTWO_Q29FL;
+			sign  = -1;
+		} else {
+			th_rad_fxp += CORDICSINE_PIOVERTWO_Q30FL;
+			sign  = 1;
+		}
+	} else {
+		sign  = 1;
+	}
+	th_rad_fxp <<= 2;
+	b_yn = 0;
+	xn = CORDICSINE_LUT_Q29FL;
+	xtmp = CORDICSINE_LUT_Q29FL;
+	ytmp = 0;
+	for (b_idx = 0; b_idx < 31; b_idx++) {
```

Comment:
should make a comment here about why this is 31 and not 32

Reply:
It's iterating over lookup table index

Reply:
exactly, it looks like you are using 32bits, you should use ARRAY_SIZE then

Reply:
Sure , thanks